mlflow/cli/genai_eval_utils.py

```diff
@@ -0,0 +1,252 @@
+
+    return resolved_scorers
+
+
+def extract_assessments_from_results(
+    results_df: pd.DataFrame, evaluation_run_id: str
+) -> list[dict[str, Any]]:
+    """
+    Extract assessments from evaluation results DataFrame.
+
+    The evaluate() function returns results with a DataFrame that contains
+    an 'assessments' column. Each row has a list of assessment dictionaries
+    with metadata including AssessmentMetadataKey.SOURCE_RUN_ID that we use to
+    filter assessments from this specific evaluation run.
+
+    Args:
+        results_df: DataFrame from evaluate() results containing assessments column
+        evaluation_run_id: The MLflow run ID from the evaluation that generated the assessments
+
+    Returns:
+        List of dictionaries with trace_id and assessments
+    """
+    output_data = []
+
+    for _, row in results_df.iterrows():
+        trace_id = row["trace_id"]
+        trace_assessments = {"trace_id": trace_id, "assessments": []}
+
+        for assessment_dict in row.get("assessments", []):
+            # Only consider assessments from the evaluation run
+            metadata = assessment_dict.get("metadata", {})
+            source_run_id = metadata.get(AssessmentMetadataKey.SOURCE_RUN_ID)
+
+            if source_run_id != evaluation_run_id:
+                continue
+
+            result_dict = {
+                "assessment_name": assessment_dict.get("assessment_name"),
+                "result": None,
+                "rationale": None,
+            }
+
+            if (feedback := assessment_dict.get("feedback")) and isinstance(feedback, dict):
+                result_dict["result"] = feedback.get("value")
+
+            if rationale := assessment_dict.get("rationale"):
+                result_dict["rationale"] = rationale
+
+            if error := assessment_dict.get("error"):
+                result_dict["error"] = str(error)
+
+            trace_assessments["assessments"].append(result_dict)
+
+        # If no assessments were found for this trace, add error markers
+        if not trace_assessments["assessments"]:
+            trace_assessments["assessments"].append(
+                {
+                    "assessment_name": NA_VALUE,
+                    "result": None,
+                    "rationale": None,
+                    "error": "No assessments found on trace",
+                }
+            )
+
+        output_data.append(trace_assessments)
+
+    return output_data
+
+
+def format_table_output(output_data: list[dict[str, Any]]) -> TableOutput:
+    """
+    Format evaluation results as table data.
+
+    Args:
+        output_data: List of trace results with assessments
+
+    Returns:
+        TableOutput dataclass containing headers and rows
+    """
+    # Extract unique assessment names from output_data to use as column headers
+    assessment_names_set = set()
+    for trace_result in output_data:
+        for assessment in trace_result["assessments"]:
```

Comment:
```suggestion
        for assessment in trace_result.get("assessments", []):
```

for safety.

Reply:
same comment to other key access that looks unsafe.

Reply:
Can't we use dataclass? `dict[key]` looks unsafe.

Reply:
Done. Also applied in other parts.